[TOPIC: GPIO] documentation refinements, test conversion, and sx1509b conversion for new API

pabigot opened the pull request:
This PR provides a few documentation refinements to the API header and converts the generic GPIO test to confirm the legacy API and new API works as intended.  It also switches the legacy API to be implemented in terms of the new API, though that may be premature.

This is based on PR #16648 under the assumption that's ready to be merged to the topic branch.

pabigot committed:
test/drivers/gpio: add tests for new API

Test that the new port API functions all behave as expected, including
physical vs logical level for input and output as well as masked and
set-based output operations.  Also tests the new pin API functions.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
gpio: document GPIO flag values in table format

Add a comment that shows the hexadecimal representation of each flag, to
help a user who needs to decode a packed set of flags.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests/drivers/gpio: replace legacy API in callback manage tests

Switch to gpio_pin_interrupt_configure() and the new interrupt flags.
Use logical level pin set operations.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
tests/drivers/gpio: replace legacy API in callback tests

Switch to gpio_pin_interrupt_configure() and the new interrupt flags.
Use logical level pin set operations.  Test all standard interrupt
configurations including double edge.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
WIP tests/drivers/gpio: generalize gpio specification solution

Provide board-specific overlays that specify which pins to use for the
test, rather than use aliases in the board device tree.  This uses a
binding that's specifically documented for use in GPIO applications.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

mnkp commented:
@pabigot Thanks for the PR.

I'm going to include the following minor updates directly into existing commits in #16648:
- gpio: document replacement for deprecated callback functions
- gpio: standardize deprecation message style
- drivers/gpio_mcux: implement initial output flags

this will keep commit history clean. The changes could be part of the review process.

I'm going to cherry pick the following commit into #16648:
- gpio: use named types to distinguish pin sets from pin values

I would like to include changes proposed in:
- gpio: replace legacy API implementation with new API

in #16648 but that needs a bit more discussion. I'll follow up.

Remaining commits can be merged later directly into topic-gpio branch.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -817,13 +823,15 @@ static inline int gpio_port_set_clr_bits_raw(struct device *port,
- * be toggled.
+ * Note that the behavior required for pins present in both
+ * ``set_pins`` and ``clear_pins`` is to toggle the current pin
+ * setting.
```
This text is written as if it was an instruction for a person implementing the driver. The API documentation should be targeted at the users of the API.

mnkp commented on include/drivers/gpio.h:
```diff
@@ -817,13 +823,15 @@ static inline int gpio_port_set_clr_bits_raw(struct device *port,
- * @param clear_pins Value indicating which pins will be set to inactive or
- *                   toggled.
+ * @param set_pins Pins selected by this mask are changed to active if
+ * they were not already active.
```
From the commit description:
> It also clarifies the expected behavior of the driver in situations
where the driver can detect that an operation would have no effect,
which is signficant for external IO extenders where an unnecessary bus
transaction can be eliminated.

This optimization can be done by the driver also without changing the description. While we mention toggle functionality in the description of the function this text doesn't make it clear that "Pins selected by this mask can be changed to inactive if a matching pin is selected by `clear_pins` parameter." I'm not saying we shouldn't change the description of `set_pins`, `clear_pins` parameters but this proposal doesn't seem like an improvement.

pabigot committed:
WIP tests/drivers/gpio: add performance tests

Time execution of a sequence of operations.

s1 test takes 7.02 us on pca10056.

pabigot committed:
drivers: gpio_nrfx: remove outdated driver functions

The access-oriented GPIO read and write system calls have been replaced
with new API.  Remove the old implementation.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
@mnkp these new commits might be worth squashing into to your branch:
* 1bcc9c1644 [Thu Aug 29 10:51:49 2019 -0500] gpio: simplify port clear bits
* c637eb44ac [Thu Aug 29 14:09:11 2019 -0500] drivers: gpio: fix handler for output parameter

mnkp commented on include/drivers/gpio.h:
```diff
@@ -779,38 +779,28 @@ static inline int z_impl_gpio_port_toggle_bits(struct device *port,
- * Pins which are simultaneously selected by set_pins and clear_pins value will
- * be toggled.
+ * Pins that are selected by both set_pins and clear_pins will be
+ * cleared to low.
```
I believe the API would be "cleaner" if the function disallowed to have a pin selected by both set_pins and clear_pins values. If the application asks to have a pin set and cleared at the same time something clearly went wrong. It's better to report an error (or catch the issue via ASSERT) rather than quietly clear the pin.

pabigot replied:
I think that'd be OK.  So the idea would be something like:

    __ASSERT((set_pins & clear_pins) == 0,
             "Pins cannot be both set and cleared");
    return gpio_port_set_masked(port, set_pins | clear_pins, set_pins);

The effect in the default situation of ignoring `__ASSERT` is that pins that are both set and cleared are actually just set, which is as good a choice as clearing them, and because this is done in the common header inline function such (illegal) usage would produce consistent behavior across drivers, which is desirable.

mnkp replied:
OK, I'll implement the change, you don't need to push an update to your PR.

pabigot committed:
drivers: gpio_mcux: remove outdated driver functions

The access-oriented GPIO read and write system calls have been replaced
with new API.  Remove the old implementation.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_gecko: remove outdated driver functions

The access-oriented GPIO read and write system calls have been replaced
with new API.  Remove the old implementation.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_sx1509b: update to use new GPIO API

Update driver code and board files to use new GPIO configuration flags
such as GPIO_ACTIVE_LOW. Also add implementation of new port_* driver
API.

Tested on external SX1509B breakout board and Thingy:52.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
DNM tests/drivers/gpio: sx1509B test infrastructure

pabigot committed:
gpio: reimplement legacy API using new API

Use the recommended replacement as the implementation of the deprecated
pin read and write functions, rather than calling the old
implementation.

Retain the existing test, but silence the complaints related to using
deprecated functions.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
gpio: remove GPIO legacy API functions

Remove the old read and write implementation functions from the driver
API table, along with their system call implementations.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
drivers: gpio_sx1509b: remove outdated driver functions

The access-oriented GPIO read and write system calls have been replaced
with new API.  Remove the old implementation.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
boards/particle_*: update for revised GPIO API

Set the devicetree nodes to correctly reflect active levels for buttons,
leds, and antenna switch.  Update the board initialization code to use
the non-deprecated GPIO API for antenna switching.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

mnkp commented on drivers/gpio/gpio_nrfx.c:
```diff
@@ -132,6 +132,12 @@ static int gpio_nrfx_config(struct device *port, int access_op,
 	u8_t from_pin;
 	u8_t to_pin;
 
+	/* Open Source is not supported on this platform */
```
There was a problem with the testcase, I needed to add pull up / pull down on a pin. I couldn't rely on the LED and resistor to set pin value. The test is passing on nrf52840_pca10056 board. The platform supports Open Drain / Open Source mode.

pabigot committed:
FIXUP gpio.h: put back deprecation comments

The comments identifying replacement API were inadvertently removed
along with the flag that triggers deprecation warnings.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
FIXUP gpio.h: add error return when blocking might occur

External GPIO drivers may not be supported from interrupt context
because they involve blocking bus transactions.  Describe the return
value for this situation, and add the I/O error to operations where it
was missing.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
Since this branch has content that won't pass shippable it's not suitable for review.  Future reviews will have to be on an unsubmitted branch.
